rfcs/1-declarative-configuration.md

```diff
@@ -0,0 +1,127 @@
+# Declarative Configuration
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details](#implementation-details)
+  - [Configuration example](#examples)
+- [Test Plan](#testrelease-plan)
+<!-- /toc -->
+
+
+## Summary
+
+This proposal replaces environment variable based configuration being used in Make+Magefiles in favor of using declarative yaml based configuration.
+
+## Motiviation
+
+As we've been writing `devbase`, and other tooling, we've identified that trying to codify all options for tooling to environment variables, bespoke configuration files, and other obscure means has made it difficult for us to provide options and maintain a common set of toolign.
+
+This, combined with a lack of documentation into all of these options, has laregely made `devbase` unconsumable outside of bootstrap/stencil. Thinking of `devbase` as it's own framework to build software makes it more usable in non-stencil projects (like we do already today w/ submodules)
+### Goals
+
+ - Documentation for all options in configuration file(s)
+ - Standard location for all configuration for tooling provided by devbase, et. al
+ - Documentation mapping 1:1 with Magefile target, with framework to consume/declare it
+ - Documentation on how to write Magefile targets, if the framework docs aren't self-explanatory enough
+ - Support for usage outside of stencil-base, et. al
+ - Not breaking, while we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal
+
+### Non-goals
+
+ - Expose new functionality, this would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.
+ - Migrating `Makefile` to `Magefile`, while it'd be nice to migrate all of them over, that's also an amount of work to do. While moving to `Magefile` would be good, we shouldn't try to move everything to Go at the same time. (TL;DR: Wrap shell were needed)
+
+## Proposal
+
+The high level proposal here is generally covered in the [Summary](#summary) and [Moviation](#motiviation) sections.
+
+### Implementation Details
+
+#### Repository Configuration
+
+Configuration would live in the root of repository consuming `devbase`, in `.devbase`.
+
+Each configuration file would match the relevant `mage <target>` target, e.g.
+
+```bash
+$ ls .devbase
+build.yaml
+e2e.yaml
+docker.yaml
+```
+
+**Note**: For shared configuration, e.g. `docker` (which could be shared between `docker-build` and `docker-push`) that's also allowed and can become a unified target.
```

Comment:
You should come up with a paradigm here. Let's take the same example with `docker-build` and `docker-push` sharing configuration in `docker.yaml`

Maybe the paradigm is this - for these targets with shared config they need to all be postfixed with `docker_` and you write shared config in the root of the yaml document, with config pertaining to that specific target under a key with the same name as what follows the prefix.

So what you'd have is actually `docker_build` and `docker_push` with a `docker.yaml` file that looks like this:
```yaml
i_am: "shared config"
here_is_more: "shared config"
build:
  more: "config"
push:
  moar: "config"
```

Reply:
the reason i did underscore instead of hyphen is because then u can still use hypen if need multiple words after the postfixed part

Reply:
Talked offline, updated this note to be more clear

Reply:
We could also nest this arbitrarily. e.g. every config file has a nested structure.


For example, `docker.yaml` would have only the config for all docker targets:
```yaml
docker:
  i-am-shared-for-docker-targets: "config"
```

Then `docker-build.yaml` would have the config for build:
```yaml
docker:
  build:
    i-am-only-used-in-build: "config"
```

That could also allow to easily overriding a config from `docker.yaml` inside `docker-build.yaml`.